File: libmmux/src/mmux/storage/client/kv_client.cpp
#include "kv_client.h"
#include "../../utils/logger.h"
#include "../../utils/string_utils.h"
#include "../kv/hash_slot.h"

namespace mmux {
namespace storage {

using namespace mmux::utils;

kv_client::kv_client(std::shared_ptr<directory::directory_ops> fs,
                     const std::string &path,
                     const directory::data_status &status,
                     int timeout_ms)
    : fs_(std::move(fs)), path_(path), status_(status) {
  slots_.clear();
  blocks_.clear();
  for (const auto &block: status.data_blocks()) {
    slots_.push_back(block.slot_begin());
    blocks_.push_back(std::make_shared<replica_chain_client>(block.block_names, timeout_ms));
  }
}

directory::data_status &kv_client::status() {
  return status_;
}

std::shared_ptr<kv_client::locked_client> kv_client::lock() {
  return std::make_shared<kv_client::locked_client>(*this);
}

void kv_client::refresh() {
  status_ = fs_->dstatus(path_);
  LOG(log_level::info) << "Refreshing block mappings to " << status_.to_string();
  slots_.clear();
  blocks_.clear();
  for (const auto &block: status_.data_blocks()) {
    slots_.push_back(block.slot_begin());
    blocks_.push_back(std::make_shared<replica_chain_client>(block.block_names));
  }
}

std::string kv_client::put(const std::string &key, const std::string &value) {
  std::string _return;
  std::vector<std::string> args{key, value};
  bool redo;
  do {
    try {
      _return = blocks_[block_id(key)]->run_command(kv_op_id::put, args).front();
      handle_redirect(kv_op_id::put, args, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::string kv_client::get(const std::string &key) {
  std::string _return;
  std::vector<std::string> args{key};
  bool redo;
  do {
    try {
      _return = blocks_[block_id(key)]->run_command(kv_op_id::get, args).front();
      handle_redirect(kv_op_id::get, args, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::string kv_client::update(const std::string &key, const std::string &value) {
  std::string _return;
  std::vector<std::string> args{key, value};
  bool redo;
  do {
    try {
      _return = blocks_[block_id(key)]->run_command(kv_op_id::update, args).front();
      handle_redirect(kv_op_id::update, args, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::string kv_client::remove(const std::string &key) {
  std::string _return;
  std::vector<std::string> args{key};
  bool redo;
  do {
    try {
      _return = blocks_[block_id(key)]->run_command(kv_op_id::remove, args).front();
      handle_redirect(kv_op_id::remove, args, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::vector<std::string> kv_client::put(const std::vector<std::string> &kvs) {
  if (kvs.size() % 2 != 0) {
    throw std::invalid_argument("Incorrect number of arguments");
  }
  std::vector<std::string> _return;
  bool redo;
  do {
    try {
      _return = batch_command(kv_op_id::put, kvs, 2);
      handle_redirects(kv_op_id::put, kvs, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::vector<std::string> kv_client::get(const std::vector<std::string> &keys) {
  std::vector<std::string> _return;
  bool redo;
  do {
    try {
      _return = batch_command(kv_op_id::get, keys, 1);
      handle_redirects(kv_op_id::get, keys, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::vector<std::string> kv_client::update(const std::vector<std::string> &kvs) {
  if (kvs.size() % 2 != 0) {
    throw std::invalid_argument("Incorrect number of arguments");
  }
  std::vector<std::string> _return;
  bool redo;
  do {
    try {
      _return = batch_command(kv_op_id::update, kvs, 2);
      handle_redirects(kv_op_id::update, kvs, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

std::vector<std::string> kv_client::remove(const std::vector<std::string> &keys) {
  std::vector<std::string> _return;
  bool redo;
  do {
    try {
      _return = batch_command(kv_op_id::remove, keys, 1);
      handle_redirects(kv_op_id::remove, keys, _return);
      redo = false;
    } catch (redo_error &e) {
      redo = true;
    }
  } while (redo);
  return _return;
}

size_t kv_client::block_id(const std::string &key) {
  return static_cast<size_t>(std::upper_bound(slots_.begin(), slots_.end(), hash_slot::get(key)) - slots_.begin() - 1);
}

std::vector<std::string> kv_client::batch_command(const kv_op_id &op,
                                                  const std::vector<std::string> &args,
                                                  size_t args_per_op) {
  // Split arguments
  if (args.size() % args_per_op != 0)
    throw std::invalid_argument("Incorrect number of arguments");

  std::vector<std::vector<std::string>> block_args(blocks_.size());
  std::vector<std::vector<size_t>> positions(blocks_.size());
  size_t num_ops = args.size() / args_per_op;
  for (size_t i = 0; i < num_ops; i++) {
    auto id = block_id(args[i * args_per_op]);
    for (size_t j = 0; j < args_per_op; j++)
      block_args[id].push_back(args[i * args_per_op + j]);
    positions[id].push_back(i);
  }

  for (size_t i = 0; i < blocks_.size(); i++) {
    if (!block_args[i].empty())
      blocks_[i]->send_command(op, block_args[i]);
  }

  std::vector<std::string> results(num_ops);
  for (size_t i = 0; i < blocks_.size(); i++) {
    if (!block_args[i].empty()) {
      auto res = blocks_[i]->recv_response();
      for (size_t j = 0; j < res.size(); j++) {
        results[positions[i][j]] = res[j];
      }
    }
  }

  return results;
}

void kv_client::handle_redirect(int32_t cmd_id, const std::vector<std::string> &args, std::string &response) {
  if (response.substr(0, 10) == "!exporting") {
    typedef std::vector<std::string> list_t;
    do {
      auto parts = string_utils::split(response, '!');
      auto chain = list_t(parts.begin() + 2, parts.end());
      response = replica_chain_client(std::move(chain)).run_command_redirected(cmd_id, args).front();
    } while (response.substr(0, 10) == "!exporting");
  }
  if (response == "!block_moved") {
    refresh();
    throw redo_error();
  }
}

void kv_client::handle_redirects(int32_t cmd_id,
                                 const std::vector<std::string> &args,
                                 std::vector<std::string> &responses) {
  size_t n_ops = responses.size();
  size_t n_op_args = args.size() / n_ops;
  for (size_t i = 0; i < responses.size(); i++) {
    auto &response = responses[i];
    if (response.substr(0, 10) == "!exporting") {
      typedef std::vector<std::string> list_t;
      list_t op_args(args.begin() + i * n_op_args, args.begin() + (i + 1) * n_op_args);
      do {
        auto parts = string_utils::split(response, '!');
        auto chain = list_t(parts.begin() + 2, parts.end());
        response = replica_chain_client(std::move(chain)).run_command_redirected(cmd_id, op_args).front();
      } while (response.substr(0, 10) == "!exporting");
    }
    if (response == "!block_moved") {
      refresh();
      throw redo_error();
    }
  }
}

kv_client::locked_client::locked_client(kv_client &parent) : parent_(parent) {
  blocks_.resize(parent_.blocks_.size());
  redirect_blocks_.resize(parent_.blocks_.size());
  locked_redirect_blocks_.resize(parent_.blocks_.size());
  new_blocks_.resize(parent_.blocks_.size());
  for (size_t i = 0; i < blocks_.size(); i++) {
    blocks_[i] = parent_.blocks_[i]->lock();
  }
  for (size_t i = 0; i < blocks_.size(); i++) {
    if (blocks_[i]->redirecting()) {
      bool new_block = true;
      for (size_t j = 0; j < blocks_.size(); j++) {
        if (blocks_[i]->redirect_chain() == blocks_[j]->chain()) {
          new_block = false;
          redirect_blocks_[i] = parent_.blocks_[j];
          locked_redirect_blocks_[i] = blocks_[j];
          new_blocks_[i] = nullptr;
          break;
        }
      }
      if (new_block) {
        redirect_blocks_[i] = std::make_shared<replica_chain_client>(blocks_[i]->chain());
        locked_redirect_blocks_[i] = redirect_blocks_[i]->lock();
        new_blocks_[i] = locked_redirect_blocks_[i];
      }
    } else {
      new_blocks_[i] = nullptr;
      redirect_blocks_[i] = nullptr;
      locked_redirect_blocks_[i] = nullptr;
    }
  }
}

void kv_client::locked_client::unlock() {
  for (size_t i = 0; i < blocks_.size(); i++) {
    blocks_[i]->unlock();
    if (new_blocks_[i] != nullptr)
      new_blocks_[i]->unlock();
  }
}

size_t kv_client::locked_client::num_keys() {
  for (size_t i = 0; i < blocks_.size(); i++) {
    blocks_[i]->send_command(kv_op_id::num_keys, {});
    if (new_blocks_[i] != nullptr) {
      new_blocks_[i]->send_command(kv_op_id::num_keys, {});
    }
  }
  size_t n = 0;
  for (size_t i = 0; i < blocks_.size(); i++) {
    n += std::stoll(blocks_[i]->recv_response().front());
    if (new_blocks_[i] != nullptr) {
      n += std::stoll(new_blocks_[i]->recv_response().front());
    }
  }
  return n;
}

std::string kv_client::locked_client::put(const std::string &key, const std::string &value) {
  std::vector<std::string> args{key, value};
  auto _return = blocks_[parent_.block_id(key)]->run_command(kv_op_id::locked_put, args).front();
  handle_redirect(kv_op_id::locked_put, args, _return);
  return _return;
}

std::string kv_client::locked_client::get(const std::string &key) {
  std::vector<std::string> args{key};
  auto _return = blocks_[parent_.block_id(key)]->run_command(kv_op_id::locked_get, args).front();
  handle_redirect(kv_op_id::locked_get, args, _return);
  return _return;
}

std::string kv_client::locked_client::update(const std::string &key, const std::string &value) {
  std::vector<std::string> args{key, value};
  auto _return = blocks_[parent_.block_id(key)]->run_command(kv_op_id::locked_update, args).front();
  handle_redirect(kv_op_id::locked_update, args, _return);
  return _return;
}

std::string kv_client::locked_client::remove(const std::string &key) {
  std::vector<std::string> args{key};
  auto _return = blocks_[parent_.block_id(key)]->run_command(kv_op_id::locked_remove, args).front();
  handle_redirect(kv_op_id::locked_remove, args, _return);
  return _return;
}

std::vector<std::string> kv_client::locked_client::put(const std::vector<std::string> &kvs) {
  auto _return = parent_.batch_command(kv_op_id::locked_put, kvs, 2);
  handle_redirects(kv_op_id::locked_put, kvs, _return);
  return _return;
}

std::vector<std::string> kv_client::locked_client::get(const std::vector<std::string> &keys) {
  auto _return = parent_.batch_command(kv_op_id::locked_get, keys, 1);
  handle_redirects(kv_op_id::locked_get, keys, _return);
  return _return;
}

std::vector<std::string> kv_client::locked_client::update(const std::vector<std::string> &kvs) {
  auto _return = parent_.batch_command(kv_op_id::locked_update, kvs, 2);
  handle_redirects(kv_op_id::locked_update, kvs, _return);
  return _return;
}

std::vector<std::string> kv_client::locked_client::remove(const std::vector<std::string> &keys) {
  auto _return = parent_.batch_command(kv_op_id::locked_remove, keys, 1);
  handle_redirects(kv_op_id::locked_remove, keys, _return);
  return _return;
}

void kv_client::locked_client::handle_redirect(int32_t cmd_id,
                                               const std::vector<std::string> &args,
                                               std::string &response) {
  if (response.substr(0, 10) == "!exporting") {
    typedef std::vector<std::string> list_t;
    do {
      auto parts = string_utils::split(response, '!');
      auto chain = list_t(parts.begin() + 2, parts.end());
      bool found = false;
      for (size_t i = 0; i < blocks_.size(); i++) {
        const auto &client_chain = parent_.blocks_[i]->chain();
        if (client_chain == chain) {
          found = true;
          response = blocks_[i]->run_command_redirected(cmd_id, args).front();
          break;
        }
      }
      if (!found)
        response = replica_chain_client(std::move(chain)).run_command_redirected(cmd_id, args).front();
    } while (response.substr(0, 10) == "!exporting");
  }
  // There can be !block_moved response, since:
  // (1) No new exports can start while the kv is locked
  // (2) Ongoing exports cannot finish while the kv is locked
}

void kv_client::locked_client::handle_redirects(int32_t cmd_id,
                                                const std::vector<std::string> &args,
                                                std::vector<std::string> &responses) {
  size_t n_ops = responses.size();
  size_t n_op_args = args.size() / n_ops;
  for (size_t i = 0; i < responses.size(); i++) {
    auto &response = responses[i];
    if (response.substr(0, 10) == "!exporting") {
      typedef std::vector<std::string> list_t;
      list_t op_args(args.begin() + i * n_op_args, args.begin() + (i + 1) * n_op_args);
      do {
        auto parts = string_utils::split(response, '!');
        auto chain = list_t(parts.begin() + 2, parts.end());
        bool found = false;
        for (size_t j = 0; j < blocks_.size(); j++) {
          const auto &client_chain = parent_.blocks_[j]->chain();
          if (client_chain == chain) {
            found = true;
            response = blocks_[j]->run_command_redirected(cmd_id, args).front();
            break;
          }
        }
        if (!found)
          response = replica_chain_client(std::move(chain)).run_command_redirected(cmd_id, op_args).front();
      } while (response.substr(0, 10) == "!exporting");
    }
  }
  // There can be !block_moved response, since:
  // (1) No new exports can start while the kv is locked
  // (2) Ongoing exports cannot finish while the kv is locked
}

}
}

File: pyclient/mmux/kv/kv_client.py
import logging
from bisect import bisect_right

from mmux.kv import crc
from mmux.kv.block_client import BlockClientCache
from mmux.kv.compat import b, unicode, bytes, long, basestring, bytes_to_str
from mmux.kv.kv_ops import KVOps
from mmux.kv.replica_chain_client import ReplicaChainClient


def encode(value):
    if isinstance(value, bytes):
        return value
    elif isinstance(value, (int, long)):
        value = b(str(value))
    elif isinstance(value, float):
        value = b(repr(value))
    elif not isinstance(value, basestring):
        value = unicode(value)
    if isinstance(value, unicode):
        value = value.encode()
    return value


class RedoError(Exception):
    pass


class RedirectError(Exception):
    def __init__(self, message, blocks):
        super(RedirectError, self).__init__(message)
        self.blocks = blocks


class KVClient:
    class LockedClient:
        def __init__(self, parent):
            self.parent = parent
            self.blocks = [block.lock() for block in self.parent.blocks]
            self.redirect_blocks = [None for _ in self.blocks]
            self.locked_redirect_blocks = [None for _ in self.blocks]
            self.new_blocks = [None for _ in self.blocks]
            for i in range(len(self.blocks)):
                if self.blocks[i].is_redirecting():
                    new_block = True
                    for j in range(len(self.blocks)):
                        if self.blocks[i].get_redirect_chain() == self.blocks[j].get_chain():
                            new_block = False
                            self.redirect_blocks[i] = self.parent.blocks[j]
                            self.locked_redirect_blocks[i] = self.blocks[j]

                    if new_block:
                        self.redirect_blocks[i] = ReplicaChainClient(self.parent.client_cache, self.blocks[i].get_chain())
                        self.locked_redirect_blocks[i] = self.redirect_blocks[i].lock()
                        self.new_blocks[i] = self.locked_redirect_blocks[i]

        def _handle_redirect(self, cmd_id, args, response):
            response = b(response)
            while response.startswith(b('!exporting')):
                chain = [bytes_to_str(x) for x in response[1:].split(b('!'))[1:]]
                found = False
                for i in range(len(self.blocks)):
                    client_chain = self.parent.blocks[i].chain()
                    if client_chain == chain:
                        found = True
                        response = self.blocks[i].run_command_redirected(cmd_id, args)[0]
                        break
                if not found:
                    response = ReplicaChainClient(self.parent.client_cache, chain).run_command_redirected(cmd_id, args)[0]
            return response

        def _handle_redirects(self, cmd_id, args, responses):
            n_ops = len(responses)
            n_op_args = int(len(args) / n_ops)
            for i in range(n_ops):
                response = b(responses[i])
                while response.startswith(b('!exporting')):
                    chain = [bytes_to_str(x) for x in response[1:].split(b('!'))[1:]]
                    op_args = args[i * n_op_args: (i + 1) * n_op_args]
                    found = False
                    for j in range(len(self.blocks)):
                        client_chain = self.parent.blocks[j].chain()
                        if client_chain == chain:
                            found = True
                            response = self.blocks[j].run_command_redirected(cmd_id, op_args)[0]
                            break
                    if not found:
                        response = ReplicaChainClient(self.parent.client_cache, chain).run_command_redirected(cmd_id, op_args)[0]
                responses[i] = response
            return responses

        def unlock(self):
            for i in range(len(self.blocks)):
                self.blocks[i].unlock()
                if self.locked_redirect_blocks[i] is not None:
                    self.locked_redirect_blocks[i].unlock()

        def put(self, key, value):
            args = [key, value]
            response = self.blocks[self.parent.block_id(key)].run_command(KVOps.locked_put, args)[0]
            return self._handle_redirect(KVOps.locked_put, args, response)

        def get(self, key):
            args = [key]
            response = self.blocks[self.parent.block_id(key)].run_command(KVOps.locked_get, args)[0]
            return self._handle_redirect(KVOps.locked_get, args, response)

        def update(self, key, value):
            args = [key, value]
            response = self.blocks[self.parent.block_id(key)].run_command(KVOps.locked_update, args)[0]
            return self._handle_redirect(KVOps.locked_update, args, response)

        def remove(self, key):
            args = [key]
            response = self.blocks[self.parent.block_id(key)].run_command(KVOps.locked_remove, args)[0]
            return self._handle_redirect(KVOps.locked_remove, args, response)

        def multi_put(self, args):
            response = self.parent._batch_command(KVOps.locked_put, args, 2)
            return self._handle_redirects(KVOps.locked_put, args, response)

        def multi_get(self, args):
            response = self.parent._batch_command(KVOps.locked_get, args, 1)
            return self._handle_redirects(KVOps.locked_get, args, response)

        def multi_update(self, args):
            response = self.parent._batch_command(KVOps.locked_update, args, 2)
            return self._handle_redirects(KVOps.locked_update, args, response)

        def multi_remove(self, args):
            response = self.parent._batch_command(KVOps.locked_remove, args, 1)
            return self._handle_redirects(KVOps.locked_remove, args, response)

        def num_keys(self):
            for i in range(len(self.blocks)):
                self.blocks[i].send_command(KVOps.num_keys, [])
                if self.new_blocks[i] is not None:
                    self.new_blocks[i].send_command(KVOps.num_keys, [])

            n = 0
            for i in range(len(self.blocks)):
                n += int(self.blocks[i].recv_response()[0])
                if self.new_blocks[i] is not None:
                    n += int(self.new_blocks[i].recv_response()[0])
            return n

    def __init__(self, fs, path, data_status, chain_failure_cb):
        self.fs = fs
        self.path = path
        self.client_cache = BlockClientCache()
        self.file_info = data_status
        self.blocks = [ReplicaChainClient(self.client_cache, chain.block_names) for chain in data_status.data_blocks]
        self.slots = [chain.slot_range[0] for chain in data_status.data_blocks]
        self.chain_failure_cb_ = chain_failure_cb

    def refresh(self):
        self.file_info = self.fs.dstatus(self.path)
        logging.info("Refreshing block mappings to {}".format(self.file_info.data_blocks))
        self.blocks = [ReplicaChainClient(self.client_cache, chain.block_names) for chain in self.file_info.data_blocks]
        self.slots = [chain.slot_range[0] for chain in self.file_info.data_blocks]

    def lock(self):
        return self.LockedClient(self)

    def _handle_redirect(self, cmd_id, args, response):
        response = b(response)
        while response.startswith(b('!exporting')):
            chain = [bytes_to_str(x) for x in response[1:].split(b('!'))[1:]]
            response = ReplicaChainClient(self.client_cache, chain).run_command_redirected(cmd_id, args)[0]
        if response == b('!block_moved'):
            self.refresh()
            return None
        return response

    def _handle_redirects(self, cmd_id, args, responses):
        n_ops = len(responses)
        n_op_args = int(len(args) / n_ops)
        for i in range(n_ops):
            response = b(responses[i])
            while response.startswith(b('!exporting')):
                chain = [bytes_to_str(x) for x in response[1:].split(b('!'))[1:]]
                op_args = args[i * n_op_args: (i + 1) * n_op_args]
                response = ReplicaChainClient(self.client_cache, chain).run_command_redirected(cmd_id, op_args)[0]
            if response == "!block_moved":
                self.refresh()
                return None
            responses[i] = response
        return responses

    def _batch_command(self, op, args, args_per_op):
        if len(args) % args_per_op != 0:
            raise RuntimeError("Incorrect number of arguments")

        num_ops = int(len(args) / args_per_op)
        block_args = [[] for _ in self.blocks]
        positions = [[] for _ in self.blocks]

        for i in range(num_ops):
            bid = self.block_id(args[i * args_per_op])
            block_args[bid].extend(args[i * args_per_op: (i + 1) * args_per_op])
            positions[bid].append(i)

        for i in range(len(self.blocks)):
            if block_args[i]:
                self.blocks[i].send_command(op, block_args[i])

        responses = [None for _ in range(num_ops)]
        for i in range(len(self.blocks)):
            if block_args[i]:
                response = self.blocks[i].recv_response()
                for j in range(len(response)):
                    responses[positions[i][j]] = response[j]

        return responses

    def put(self, key, value):
        args = [key, value]
        response = None
        while response is None:
            response = self.blocks[self.block_id(key)].run_command(KVOps.put, args)[0]
            response = self._handle_redirect(KVOps.put, args, response)
        return response

    def get(self, key):
        args = [key]
        response = None
        while response is None:
            response = self.blocks[self.block_id(key)].run_command(KVOps.get, args)[0]
            response = self._handle_redirect(KVOps.get, args, response)
        return response

    def exists(self, key):
        args = [key]
        response = None
        while response is None:
            response = self.blocks[self.block_id(key)].run_command(KVOps.exists, args)[0]
            response = self._handle_redirect(KVOps.exists, args, response)
        return response == b'true'

    def update(self, key, value):
        args = [key, value]
        response = None
        while response is None:
            response = self.blocks[self.block_id(key)].run_command(KVOps.update, args)[0]
            response = self._handle_redirect(KVOps.update, args, response)
        return response

    def remove(self, key):
        args = [key]
        response = None
        while response is None:
            response = self.blocks[self.block_id(key)].run_command(KVOps.remove, args)[0]
            response = self._handle_redirect(KVOps.remove, args, response)
        return response

    def multi_put(self, args):
        response = None
        while response is None:
            response = self._batch_command(KVOps.put, args, 2)
            response = self._handle_redirects(KVOps.put, args, response)
        return response

    def multi_get(self, args):
        response = None
        while response is None:
            response = self._batch_command(KVOps.get, args, 1)
            response = self._handle_redirects(KVOps.get, args, response)
        return response

    def multi_exists(self, args):
        response = None
        while response is None:
            response = self._batch_command(KVOps.exists, args, 1)
            response = self._handle_redirects(KVOps.exists, args, response)
        return [r == b'true' for r in response]

    def multi_update(self, args):
        response = None
        while response is None:
            response = self._batch_command(KVOps.update, args, 2)
            response = self._handle_redirects(KVOps.update, args, response)
        return response

    def multi_remove(self, args):
        response = None
        while response is None:
            response = self._batch_command(KVOps.remove, args, 1)
            response = self._handle_redirects(KVOps.remove, args, response)
        return response

    def block_id(self, key):
        i = bisect_right(self.slots, crc.crc16(encode(key)))
        if i:
            return i - 1
        raise ValueError
